x-pack/plugins/reporting/server/export_types/common/pdf/pdfmaker.ts

```diff
@@ -5,52 +5,73 @@
 import { getFont } from './get_font';
-import { getTemplate } from './get_template';
-
-const assetPath = path.resolve(__dirname, '..', '..', 'common', 'assets');
-const tableBorderWidth = 1;
+import {
+  headingHeight,
+  pageMarginBottom,
+  pageMarginTop,
+  pageMarginWidth,
+  subheadingHeight,
+  tableBorderWidth,
+} from './constants';
+import { PdfWorkerOutOfMemoryError } from './pdfmaker_errors';
+
+import type { GeneratePdfRequest, GeneratePdfResponse } from './worker';
+// Ensure that the worker and all dependencies are included in the release bundle.
+import './worker';
 
 export class PdfMaker {
-  private _layout: Layout;
+  _layout: Layout;
   private _logo: string | undefined;
   private _title: string;
   private _content: Content[];
-  private _printer: Printer;
-  private _pdfDoc: PDFKit.PDFDocument | undefined;
 
-  constructor(layout: Layout, logo: string | undefined) {
-    const fontPath = (filename: string) => path.resolve(assetPath, 'fonts', filename);
-    const fonts = {
-      Roboto: {
-        normal: fontPath('roboto/Roboto-Regular.ttf'),
-        bold: fontPath('roboto/Roboto-Medium.ttf'),
-        italics: fontPath('roboto/Roboto-Italic.ttf'),
-        bolditalics: fontPath('roboto/Roboto-Italic.ttf'),
-      },
-      'noto-cjk': {
-        // Roboto does not support CJK characters, so we'll fall back on this font if we detect them.
-        normal: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
-        bold: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
-        italics: fontPath('noto/NotoSansCJKtc-Regular.ttf'),
-        bolditalics: fontPath('noto/NotoSansCJKtc-Medium.ttf'),
-      },
-    };
+  private worker?: Worker;
+
+  protected workerModulePath = path.resolve(__dirname, './worker.js');
+
+  /**
+   * The maximum heap size for old memory region of the worker thread.
+   *
+   * @note We need to provide a sane number given that we need to load a
+   * node environment for TS compilation, some library code and buffers that
+   * result from generating a PDF.
+   *
+   * Local testing indicates that to trigger an OOM event for the worker we need
+   * to exhaust not only heap but also any compression optimization and fallback
+   * swap space.
+   *
+   * With this value we are able to generate PDFs in excess of 5000x5000 pixels
+   * at which point issues other than memory start to show like glitches in the
+   * image.
+   *
+   * TODO: Should we consider making this number dynamic? It is difficult to
+   * know how we'd describe this to users in the most helpful way.
+   */
+  protected workerMaxOldHeapSizeMb = 128;
+
+  /**
+   * The maximum heap size for young memory region of the worker thread.
+   *
+   * @note leave 'undefined' to use the Node.js default value.
+   * @note we set this to a low value to trigger an OOM event sooner for the worker
+   * in test scenarios.
+   */
+  protected workerMaxYoungHeapSizeMb: number | undefined = undefined;
```

Comment:
Maybe we should just ignore the existence of this field? Not passing anything will use the Node.js default value.

Reply:
Right, the main reason I wanted to keep this is for testing convenience, it's interesting that you never hit this limit in your local testing. I think that speaks to the variability with which node applies resource limits across host machines.

The intention is to always leave it unset except for under test conditions. Is this an acceptable way of doing that? We could refactor to pass these values in through the constructor, but I through making them `protected` would be a good way of keeping the PDFMaker API clear of any worker-related settings that I don't expect we should be getting from the outside world. Let me know what you think!

Reply:
Thanks, doing it this way makes sense.